bots/src/accumulator.rs

```diff
@@ -0,0 +1,170 @@
+        Self {
+            client,
+            signer,
+            market,
+            timeout,
+        }
+    }
+
+    fn create_accumulate_tx(
+        &self,
+        borrow: &AccountId,
+        nonce: u64,
+        block_hash: CryptoHash,
+    ) -> Transaction {
+        Transaction::V0(TransactionV0 {
+            nonce,
+            receiver_id: self.market.clone(),
+            block_hash,
+            signer_id: self.signer.account_id.clone(),
+            public_key: self.signer.public_key().clone(),
+            actions: vec![Action::FunctionCall(Box::new(FunctionCallAction {
+                method_name: "apply_interest".to_string(),
+                args: serialize_and_encode(json!({
+                    "account_id": borrow,
+                })),
+                gas: DEFAULT_GAS,
+                deposit: 0,
+            }))],
+        })
+    }
+
+    #[instrument(skip(self), level = "debug")]
+    pub async fn accumulate(&self, borrow: AccountId) -> anyhow::Result<()> {
+        info!("Starting accumulation for market: {}", self.market);
+
+        let (nonce, block_hash) = get_access_key_data(&self.client, &self.signer).await?;
+
+        let apply_interest_tx = self.create_accumulate_tx(&borrow, nonce, block_hash);
+
+        match send_tx(&self.client, &self.signer, self.timeout, apply_interest_tx).await {
+            Ok(_) => {
+                info!("Accumulation successful");
+            }
+            Err(e) => {
+                error!("Accumulation failed: {e}");
+            }
+        }
+
+        Ok(())
+    }
+
+    #[instrument(skip(self), level = "debug")]
+    async fn get_borrows(&self) -> anyhow::Result<BorrowPositions> {
+        let mut all_positions: BorrowPositions = HashMap::new();
+
+        let page_size = 100;
+        let mut current_offset = 0;
+        let mut params = json!({
+            "offset": current_offset,
+            "count": page_size,
+        });
+
+        while let Ok(page) = view::<BorrowPositions>(
+            &self.client,
+            self.market.clone(),
+            "list_borrow_positions",
+            params.clone(),
+        )
+        .await
+        {
+            let fetched = page.len();
+            all_positions.extend(page);
+            current_offset += page_size;
+            params["offset"] = current_offset.into();
+
+            if fetched < page_size {
+                break;
+            }
+        }
+
+        Ok(all_positions)
+    }
+
+    #[instrument(skip(self), level = "info")]
+    pub async fn run_accumulations(&self, concurrency: usize) -> anyhow::Result<()> {
+        let borrows = self.get_borrows().await?;
+
+        futures::stream::iter(borrows)
+            .map(|(account_id, _)| async move { self.accumulate(account_id).await })
+            .buffer_unordered(concurrency)
+            .collect::<Vec<_>>()
+            .await
+            .into_iter()
+            .collect::<anyhow::Result<Vec<_>>>()?;
+
+        Ok(())
+    }
```

Comment:
Noticed a double collect here, and added an early exit.

```suggestion
    #[instrument(skip(self), level = "info")]
    pub async fn run_accumulations(&self, concurrency: usize) -> anyhow::Result<()> {
        let borrows = self.get_borrows().await?;
        
        if borrows.is_empty() {
            return Ok(());
        }

        futures::stream::iter(borrows)
            .map(|(account_id, _)| async move { self.accumulate(account_id).await })
            .buffer_unordered(concurrency)
            .try_for_each(|_result| async { Ok(()) })
            .await?;

        Ok(())
    }
```